File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
// 1012

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int M, N, K, x, y;

int bfs(vector <vector <int>> &nodes, vector <vector <bool>> &visited, int i, int j)
{
    int now_y, now_x;
    int cnt = 0;
    queue <pair <int, int>> q;
    if (!visited[i][j] && nodes[i][j] == 1)
    {
        cnt ++;
        visited[i][j] = true;
        q.push({i, j});
        while (!q.empty())
        {
            now_y = q.front().first;
            now_x = q.front().second;
            q.pop();
            if (now_y + 1 < N && !visited[now_y + 1][now_x] && nodes[now_y + 1][now_x] == 1)
            {
                visited[now_y + 1][now_x] = true;
                q.push({now_y + 1, now_x});
            }
            if (now_y - 1 >= 0 && !visited[now_y - 1][now_x] && nodes[now_y - 1][now_x] == 1)
            {
                visited[now_y - 1][now_x] = true;
                q.push({now_y - 1, now_x});
            }
            if (now_x + 1 < M && !visited[now_y][now_x + 1] && nodes[now_y][now_x + 1] == 1)
            {
                visited[now_y][now_x + 1] = true;
                q.push({now_y, now_x + 1});
            }
            if (now_x - 1 >= 0 && !visited[now_y][now_x - 1] && nodes[now_y][now_x - 1] == 1)
            {
                visited[now_y][now_x - 1] = true;
                q.push({now_y, now_x - 1});
            }
        }
        return (cnt);
    }
    return (0);
}

int main(void)
{
    int T;

    cin.tie(0);
    ios::sync_with_stdio(0);

    cin >> T;
    for (int t = 0; t < T; t ++)
    {
        int res = 0;
        cin >> M >> N >> K;
        vector <vector <int>> nodes(N, vector <int> (M, 0));
        vector <vector <bool>> visited(N, vector <bool> (M, false));
        for (int j = 0; j < K; j ++)
        {
            cin >> x >> y;
            nodes[y][x] = 1;
        }
        for (int i = 0; i < N; i ++)
        {
            for (int j = 0; j < M; j ++)
            {
                res = res + bfs(nodes, visited, i, j);
            }
        }
        cout << res << "\n";
    }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.util.*;

public class Main {
    static int M, N, K;
    static int[][] nodes;
    static boolean[][] visited;

    public static int bfs(int i, int j) {
        if (!visited[i][j] && nodes[i][j] == 1) {
            Queue<int[]> q = new LinkedList<>();
            visited[i][j] = true;
            q.offer(new int[]{i, j});

            while (!q.isEmpty()) {
                int[] current = q.poll();
                int nowY = current[0];
                int nowX = current[1];

                int[] dy = {1, -1, 0, 0};
                int[] dx = {0, 0, 1, -1};

                for (int dir = 0; dir < 4; dir++) {
                    int ny = nowY + dy[dir];
                    int nx = nowX + dx[dir];

                    if (ny >= 0 && ny < N && nx >= 0 && nx < M) {
                        if (!visited[ny][nx] && nodes[ny][nx] == 1) {
                            visited[ny][nx] = true;
                            q.offer(new int[]{ny, nx});
                        }
                    }
                }
            }
            return 1;
        }
        return 0;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int T = sc.nextInt();

        for (int t = 0; t < T; t++) {
            M = sc.nextInt();
            N = sc.nextInt();
            K = sc.nextInt();

            nodes = new int[N][M];
            visited = new boolean[N][M];

            for (int i = 0; i < K; i++) {
                int x = sc.nextInt();
                int y = sc.nextInt();
                nodes[y][x] = 1;
            }

            int res = 0;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < M; j++) {
                    res += bfs(i, j);
                }
            }
            System.out.println(res);
        }
        sc.close();
    }
}